Sources/PackageCollections/Storage/PackageCollectionsStorage.swift

```diff
@@ -0,0 +1,321 @@
+
+    /// Returns `PackageCollections` for the given identifiers, or all if non specified.
+    ///
+    /// - Parameters:
+    ///   - identifiers: Optional. The identifiers of the `PackageCollection`
+    ///   - callback: The closure to invoke when result becomes available
+    func list(identifiers: [PackageCollectionsModel.PackageCollectionIdentifier]?,
+              callback: @escaping (Result<[PackageCollectionsModel.PackageCollection], Error>) -> Void)
+
+    /// Returns `PackageSearchResult` for the given search criteria, or all if non specified.
+    ///
+    /// - Parameters:
+    ///   - identifiers: Optional. The identifiers of the `PackageCollection`
+    ///   - query: The search query expression
+    ///   - fields: Optional. Fields to search on
+    ///   - callback: The closure to invoke when result becomes available
+    func searchPacakges(identifiers: [PackageCollectionsModel.PackageCollectionIdentifier]?,
+                        query: String,
+                        callback: @escaping (Result<PackageCollectionsModel.PackageSearchResult, Error>) -> Void)
+
+    /// Returns `TargetSearchResult` for the given search criteria, or all if non specified.
+    ///
+    /// - Parameters:
+    ///   - identifiers: Optional. The identifiers of the `PackageCollection`
+    ///   - query: The search query expression
+    ///   - type: The search type
+    ///   - callback: The closure to invoke when result becomes available
+    func searchTargets(identifiers: [PackageCollectionsModel.PackageCollectionIdentifier]?,
+                       query: String,
+                       type: PackageCollectionsModel.TargetSearchType,
+                       callback: @escaping (Result<PackageCollectionsModel.TargetSearchResult, Error>) -> Void)
+}
+
+// MARK: - SQLitePackageCollectionsStorage
+
+final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable {
+    let fileSystem: FileSystem
+    let location: SQLite.Location
+
+    private let jsonEncoder: JSONEncoder
+    private let jsonDecoder: JSONDecoder
+
+    // for concurrent for DB access
+    private let queue = DispatchQueue(label: "org.swift.swiftpm.SQLitePackageCollectionsStorage", attributes: .concurrent)
+
+    private var state = State.idle
+    private let stateLock = Lock()
+
+    init(location: SQLite.Location? = nil) {
+        self.location = location ?? .path(Paths.cache(fileSystem: localFileSystem).appending(components: "package-collection.db"))
+        switch self.location {
+        case .path, .temporary:
+            self.fileSystem = localFileSystem
+        case .memory:
+            self.fileSystem = InMemoryFileSystem()
+        }
+        self.jsonEncoder = JSONEncoder()
+        self.jsonDecoder = JSONDecoder()
+    }
+
+    convenience init(path: AbsolutePath) {
+        self.init(location: .path(path))
+    }
+
+    deinit {
+        // FIXME: diagnostics instead?
```

Comment:
👀

Reply:
A fatal error seems unfortunate here (especially when the code is running the same process as the client, e.g. an IDE).  But I'm not sure whether this should be a user-facing diagnostic, since it represent a programming error.  Also, the `deinit` doesn't really provide any kind of scope in which to present that diagnostic, e.g. it's not clear how to associate it with a user action.

Since this is a programming error, I think it might make sense to add some facility to TSC to report it.  In Xcode there is a way to show a warning once on console, asking end users to file a bug report about it.  We might consider something like that for TSC, to warn and then do best-effort to recover.

That should probably be a separate PR, so I don't know what the right thing to do is here.  Maybe put the fatalError inside a #if DEBUG, since it seems unwanted in shipping code.

Reply:
I think an assertion could make sense instead since those won't trigger in release builds anyway.

Reply:
That makes sense to me.

Reply:
:+1: will change to assertion